### ISSUE TITLE
Support MySQL socket connections

### ISSUE THREAD

**Issue by bedeoverend** [metabase/metabase#3852]
Be great if metabase could support MySQL socket connections - I know in majority of cases, TCP can be used instead. But for example in my case, I'm trying to run metabase in Google App Engine and have it talk to Cloud SQL, which can only be done via a unix socket.

Ideally should be handled for both connecting to MySQL DBs, and also as the DB to store metabase's configuration on startup.

**Comment by shaundubuque** [comment]
+1 Same issue running in Google Container Engine.

**Comment by camsaul** [comment]
@bedeoverend @shaundubuque if either of you have a sense of how to accomplish this in Java or Clojure, we'd love to have you as a contributor and would welcome a PR to support this 👍

**Comment by bedeoverend** [comment]
@camsaul Unfortunately never used Clojure - have a small amount of Java experience, if I can, will try make a PR. Would love to help out - metabase is fantastic :)

**Comment by laixer** [comment]
Unfortunately Java doesn't support UNIX sockets out of the box.
For Cloud SQL, we provide a custom socket library that handles connectivity by instance name: https://github.com/GoogleCloudPlatform/cloud-sql-mysql-socket-factory

Using it in Java requires setting the `socketFactory` and the custom `cloudSqlInstance` properties, but it doesn't look like metabase provides a way to set any custom properties in the JDBC URL.

**Comment by camsaul** [comment]
@laixer we added a way to set custom JDBC connection string options in our latest release:

![screen shot 2016-12-13 at 12 37 47 pm](https://cloud.githubusercontent.com/assets/1455846/21158182/0b56e6a0-c131-11e6-939a-d59cdf92317b.png)

So you can certainly set properties like `socketFactory` and `cloudSqlInstance.

**Comment by bedeoverend** [comment]
@camsaul is this possible for the application database?

**Comment by camsaul** [comment]
@bedeoverend we do support a `MB_DB_CONNECTION_URI` environment variable so that might work if you pass in a JDBC connection URL for your MySQL database. 

We added it primarily to support Heroku so I'm not 100% sure that it will work for your case and pass along the connection string parameters correctly but give it a try and let me know if it does. If it's not passing them along we can certainly tweak it a bit.

**Comment by bedeoverend** [comment]
@camsaul thanks I'll give that a shot and report back here

**Comment by bedeoverend** [comment]
Also thanks @laixer for looking into how to get this working in App Engine

**Comment by jaredlt** [comment]
Should I be able to set `Additional JDBC connection string options` to `cloudSqlInstance=<INSTANCE_CONNECTION_NAME>&socketFactory=com.google.cloud.sql.mysql.SocketFactory` and get this to work? Or are we still waiting on Metabase to implement something?

My struggle documented here: https://stackoverflow.com/questions/59787457/how-to-connect-to-a-google-cloud-sql-database-from-metabase-when-metabase-is-ru

**Comment by ctivanovich** [comment]
I am running into this issue as well. Two years have elapsed and this is still open, so I am guessing it has not been solved. I've tried several permutations of jdbc:mysql to no avail, to connect to a cloud sql db as a data source while running metabase on google app engine.

**Comment by camsaul** [comment]
So this will work for the application DB now that #19961 is merged in. For data warehouse DBs we'd have to tweak things a bit so host and port aren't required. An example JDBC connection string would be

```
jdbc:mysql:///<DATABASE_NAME>?unixSocketPath=</PATH/TO/UNIX/SOCKET>&cloudSqlInstance=<INSTANCE_CONNECTION_NAME>&socketFactory=com.google.cloud.sql.mysql.SocketFactory&user=<MYSQL_USER_NAME>&password=<MYSQL_USER_PASSWORD>
```

(from https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/docs/jdbc-mysql.md)

e.g.

https://github.com/metabase/metabase/pull/19302/files#diff-9f5009acaa163d14003172ae09534315aecfc44149a159c8a078999cdfc287d7R28

**Comment by camsaul** [comment]
I guess we'd actually also have to not assume `localhost` is the default if `:host` isn't specified -- it happens in https://github.com/metabase/metabase/blob/master/src/metabase/db/spec.clj#L43-L51. Maybe the right answer is to not merge in the default host if `socketFactory` is specified

**Comment by camsaul** [comment]
So this issue is really asking for a few separate things 

- The general issue is to support connection options in a way that would let you specific socket connections. For the application database, this was already fixed by #19961. For data warehouse databases, we will be tracking the same project in #27116 going forward
- The more specific issue that is to support GCP socket connections, which requires shipping the Google cloud socket factory. This is being tracked in #27117

I'm closing this out in favor of those two issues going forward
